File: App/backend/app/services/notification_service.py
"""Bildirim servisi — Kullanıcı gezinme loglarına dayanarak AI tabanlı
kişiselleştirilmiş uygulama içi bildirimler üretir.

Günde 1 kez çalışır. Kullanıcının en çok ilgi gösterdiği kategori ve
bölge bilgisine göre bildirim mesajı oluşturur.
"""
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy.future import select
from sqlalchemy import func, distinct, desc, and_
from datetime import datetime, timedelta
from uuid import UUID
from typing import Optional, List

from .. import models
from .analytics_service import get_user_category_interests, get_district_category_trends


# Kategori bazlı bildirim şablonları
NOTIFICATION_TEMPLATES = {
    "sports": {
        "title": "🏃 Spor etkinlikleri ilgini çekiyor!",
        "body": "{district} bölgesinde spor etkinlikleri çok ilgi görüyor. Bir spor etkinliği planlamak ister misin?",
    },
    "music": {
        "title": "🎵 Müzik etkinlikleri seni bekliyor!",
        "body": "{district} bölgesinde müzik etkinlikleri popüler. Yakınındaki konserleri keşfetmeye ne dersin?",
    },
    "community": {
        "title": "👥 Topluluk etkinlikleri revaçta!",
        "body": "{district} bölgesinde topluluk etkinlikleri çok ilgi görüyor. Sen de bir topluluk etkinliği düzenle!",
    },
    "technology": {
        "title": "💻 Teknoloji buluşmaları popüler!",
        "body": "{district} bölgesinde teknoloji etkinlikleri trend. Bir tech meetup planlamak ister misin?",
    },
    "workshop": {
        "title": "🎨 Atölye etkinlikleri ilgi çekiyor!",
        "body": "{district} bölgesinde atölye etkinlikleri popüler. Yaratıcı bir atölye düzenle!",
    },
    "concert": {
        "title": "🎤 Konser etkinlikleri çok sevildi!",
        "body": "{district} bölgesinde konser etkinlikleri büyük ilgi görüyor. Yakınındaki konserleri kaçırma!",
    },
}

DEFAULT_TEMPLATE = {
    "title": "✨ İlgini çekebilecek etkinlikler var!",
    "body": "{district} bölgesinde {category} etkinlikleri çok ilgi görüyor. Keşfetmeye ne dersin?",
}


async def has_notification_today(db: AsyncSession, user_id: UUID) -> bool:
    """Kullanıcıya bugün zaten bildirim gönderilmiş mi kontrol eder."""
    today_start = datetime.utcnow().replace(hour=0, minute=0, second=0, microsecond=0)
    result = await db.execute(
        select(func.count(models.Notification.id))
        .where(
            models.Notification.user_id == user_id,
            models.Notification.created_at >= today_start,
        )
    )
    count = result.scalar() or 0
    return count > 0


async def generate_daily_notification(
    db: AsyncSession,
    user_id: UUID,
) -> Optional[models.Notification]:
    """Kullanıcı için günlük AI tabanlı bildirim üretir.
    
    1. Kullanıcının en çok ilgilendiği kategoriyi bulur
    2. O kategorinin en popüler olduğu bölgeyi bulur
    3. Kişiselleştirilmiş bildirim mesajı oluşturur
    
    Bugün zaten bildirim varsa None döner.
    """
    # Bugün zaten bildirim gönderilmiş mi?
    if await has_notification_today(db, user_id):
        return None

    # Kullanıcının kategori ilgilerini al
    interests = await get_user_category_interests(db, user_id)
    if not interests:
        return None

    top_category = interests[0]["category"]

    # Bu kategorinin en popüler olduğu bölgeyi bul
    top_district = await _get_top_district_for_category(db, top_category)
    district_name = top_district or "Çevren"

    # Şablondan bildirim mesajı oluştur
    template = NOTIFICATION_TEMPLATES.get(top_category, DEFAULT_TEMPLATE)
    title = template["title"]
    body = template["body"].format(
        district=district_name,
        category=top_category,
    )

    # Bildirimi kaydet
    notification = models.Notification(
        user_id=user_id,
        title=title,
        body=body,
        notification_type="event_suggestion",
        related_category=top_category,
        related_district=district_name,
    )
    db.add(notification)
    await db.commit()
    await db.refresh(notification)
    return notification


async def _get_top_district_for_category(
    db: AsyncSession,
    category: str,
    days: int = 30,
) -> Optional[str]:
    """Belirli bir kategorinin en popüler olduğu bölgeyi döner."""
    since = datetime.utcnow() - timedelta(days=days)
    result = await db.execute(
        select(
            models.EventBrowsingLog.event_location_name,
            func.count(models.EventBrowsingLog.id).label("cnt"),
        )
        .where(
            models.EventBrowsingLog.event_category == category,
            models.EventBrowsingLog.event_location_name.isnot(None),
            models.EventBrowsingLog.created_at >= since,
        )
        .group_by(models.EventBrowsingLog.event_location_name)
        .order_by(desc("cnt"))
        .limit(1)
    )
    row = result.first()
    return row.event_location_name if row else None


async def get_user_notifications(
    db: AsyncSession,
    user_id: UUID,
    limit: int = 20,
    unread_only: bool = False,
) -> List[models.Notification]:
    """Kullanıcının bildirimlerini listeler."""
    query = (
        select(models.Notification)
        .where(models.Notification.user_id == user_id)
        .order_by(desc(models.Notification.created_at))
        .limit(limit)
    )
    if unread_only:
        query = query.where(models.Notification.is_read == False)

    result = await db.execute(query)
    return result.scalars().all()


async def mark_notification_read(
    db: AsyncSession,
    notification_id: UUID,
    user_id: UUID,
) -> Optional[models.Notification]:
    """Bildirimi okundu olarak işaretler."""
    result = await db.execute(
        select(models.Notification).where(
            models.Notification.id == notification_id,
            models.Notification.user_id == user_id,
        )
    )
    notification = result.scalar_one_or_none()
    if notification:
        notification.is_read = True
        await db.commit()
        await db.refresh(notification)
    return notification

File: App/mobile/lib/services/notification_service.dart
/// Bildirim servisi — Backend'den AI tabanlı bildirimleri alır ve yönetir.
///
/// Uygulama açıldığında günlük bildirim üretimini tetikler (günde 1 kez).
/// Mock modda örnek bildirimler gösterir.
import 'api_client.dart';

class AppNotification {
  final String id;
  final String title;
  final String body;
  final String notificationType;
  final String? relatedCategory;
  final String? relatedDistrict;
  final bool isRead;
  final DateTime createdAt;

  AppNotification({
    required this.id,
    required this.title,
    required this.body,
    required this.notificationType,
    this.relatedCategory,
    this.relatedDistrict,
    required this.isRead,
    required this.createdAt,
  });

  factory AppNotification.fromJson(Map<String, dynamic> json) {
    return AppNotification(
      id: json['id'],
      title: json['title'],
      body: json['body'],
      notificationType: json['notification_type'],
      relatedCategory: json['related_category'],
      relatedDistrict: json['related_district'],
      isRead: json['is_read'] ?? false,
      createdAt: DateTime.parse(json['created_at']),
    );
  }
}

class NotificationService {
  static final NotificationService _instance = NotificationService._internal();
  factory NotificationService() => _instance;
  NotificationService._internal();

  final ApiClient _api = ApiClient();

  /// Mock bildirimler — backend bağlantısı olmadan UI'ı test etmek için.
  static final List<AppNotification> _mockNotifications = [
    AppNotification(
      id: 'mock-1',
      title: '🏃 Spor etkinlikleri ilgini çekiyor!',
      body: 'Riverside Park bölgesinde spor etkinlikleri çok ilgi görüyor. Bir spor etkinliği planlamak ister misin?',
      notificationType: 'event_suggestion',
      relatedCategory: 'sports',
      relatedDistrict: 'Riverside Park',
      isRead: false,
      createdAt: DateTime.now().subtract(const Duration(hours: 2)),
    ),
    AppNotification(
      id: 'mock-2',
      title: '🎵 Müzik etkinlikleri seni bekliyor!',
      body: 'Downtown bölgesinde müzik etkinlikleri popüler. Yakınındaki konserleri keşfetmeye ne dersin?',
      notificationType: 'event_suggestion',
      relatedCategory: 'music',
      relatedDistrict: 'Downtown',
      isRead: true,
      createdAt: DateTime.now().subtract(const Duration(days: 1)),
    ),
    AppNotification(
      id: 'mock-3',
      title: '💻 Teknoloji buluşmaları popüler!',
      body: 'Tech District bölgesinde teknoloji etkinlikleri trend. Bir tech meetup planlamak ister misin?',
      notificationType: 'create_event_prompt',
      relatedCategory: 'technology',
      relatedDistrict: 'Tech District',
      isRead: true,
      createdAt: DateTime.now().subtract(const Duration(days: 2)),
    ),
  ];

  /// Bildirimleri getirir. Mock modda örnek bildirimler döner.
  Future<List<AppNotification>> getNotifications() async {
    if (_api.useMock) return _mockNotifications;

    final data = await _api.getList('/notifications');
    if (data == null) return [];
    return data.map((j) => AppNotification.fromJson(j)).toList();
  }

  /// Günlük bildirim üretimini tetikler (uygulama açılışında çağrılır).
  Future<AppNotification?> triggerDailyNotification() async {
    if (_api.useMock) return null; // Mock modda otomatik üretme

    final data = await _api.post('/notifications/generate', {});
    if (data == null || data['status'] == 'mock_ok') return null;
    return AppNotification.fromJson(data);
  }

  /// Bildirimi okundu olarak işaretler.
  Future<void> markAsRead(String notificationId) async {
    if (_api.useMock) {
      // Mock modda lokal olarak güncelle
      final idx = _mockNotifications.indexWhere((n) => n.id == notificationId);
      if (idx >= 0) {
        _mockNotifications[idx] = AppNotification(
          id: _mockNotifications[idx].id,
          title: _mockNotifications[idx].title,
          body: _mockNotifications[idx].body,
          notificationType: _mockNotifications[idx].notificationType,
          relatedCategory: _mockNotifications[idx].relatedCategory,
          relatedDistrict: _mockNotifications[idx].relatedDistrict,
          isRead: true,
          createdAt: _mockNotifications[idx].createdAt,
        );
      }
      return;
    }

    await _api.patch('/notifications/$notificationId');
  }

  /// Okunmamış bildirim sayısını döner.
  Future<int> getUnreadCount() async {
    final notifications = await getNotifications();
    return notifications.where((n) => !n.isRead).length;
  }
}
